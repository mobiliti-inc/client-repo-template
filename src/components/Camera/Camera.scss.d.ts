// This file is automatically generated.
// Please do not change this file!
interface CssExports {
  'bottom': string;
  'bottom-left': string;
  'bottom-right': string;
  'button': string;
  'canvas': string;
  'container': string;
  'fades-in-out': string;
  'hide': string;
  'markers': string;
  'markers-ctn': string;
  'top': string;
  'top-left': string;
  'top-right': string;
  'video': string;
}
declare var cssExports: CssExports;
export = cssExports;
