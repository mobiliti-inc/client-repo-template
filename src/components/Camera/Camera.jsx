import React from 'react';
import styles from './Camera.scss';

class Camera extends React.PureComponent {
	static propTypes = {
		takePhoto: PropTypes.bool.isRequired,
		height: PropTypes.number.isRequired,
		width: PropTypes.number.isRequired,
		onTakingPhoto: PropTypes.func.isRequired,
		closeCamera: PropTypes.bool.isRequired,
		onCameraAccessFail: PropTypes.func,
		onCameraAccessSuccess: PropTypes.func,

	}

	static defaultProps = {
		onCameraAccessSuccess: () => {},
		onCameraAccessFail: () => {}
	};

	constructor(props) {
		super(props);
		this.videoStream = React.createRef();
		this.canvas = React.createRef();
		this.state = {
			photoTaken: false,
			localStream: null,
			wait: false
		};

		this.constraints = {
			video: true
		};
		this.timer = () => {};
	}
	componentWillMount() {
		if (this.hasGetUserMedia()) {
			this.initiateStream();
		} else {
			const error = {
				error: 'Unable to access user media'
			};
			this.props.onCameraAccessFail(error);
		}
	}

	componentDidUpdate(prevProps) {
		if (this.props.closeCamera !== prevProps.closeCamera && !this.state.photoTaken) {
			this.stopTracks();
		}
		if (this.props.takePhoto !== prevProps.takePhoto && this.props.takePhoto) {
			if (this.state.photoTaken && !this.props.closeCamera) {
				this.retake();
			}
		}

		if ((this.props.takePhoto !== prevProps.takePhoto) &&
		(!this.state.photoTaken || !this.props.takePhoto)) {
			this.captureImage();
		}
	}

	componentWillUnmount = () => {
		clearTimeout(this.timer);
		this.stopTracks();
	}

	gotStream = (stream) => {
		if (this.videoStream.current) {
			this.videoStream.current.srcObject = stream;
			this.setState({
				localStream: stream
			});
			this.props.onCameraAccessSuccess();
		}
		this.timer = setTimeout(() => this.setState({ wait: true }), 100);
	}

	initiateStream = () => {
		navigator.mediaDevices.getUserMedia(this.constraints)
			.then((stream) => this.gotStream(stream)).catch(this.handleError);
	}

	hasGetUserMedia = () => !!(navigator.mediaDevices &&
		navigator.mediaDevices.getUserMedia);

	handleError = (error) => {
		this.props.onCameraAccessFail(error);
	}

	captureImage = () => {
		this.setState({
			photoTaken: true
		});
		const context = this.canvas.current.getContext('2d');
		context.imageSmoothingEnabled = false;
		context.height = this.props.height;
		context.width = this.props.width;
		context.drawImage(this.videoStream.current, 0, 0, this.props.width, this.props.height);
		const image = this.canvas.current.toDataURL('image/jpeg', 0.5);
		this.stopTracks();
		return this.props.onTakingPhoto(image);
	}

	stopTracks = () => {
		const { localStream } = this.state;
		if (localStream) {
			const tracks = this.state.localStream.getTracks();
			this.videoStream.current.srcObject = null;
			tracks.forEach((track) => {
				track.stop();
			});
		}
	}

	retake = () => {
		this.setState({
			photoTaken: false,
			wait: false
		});
		this.initiateStream();
	}

	renderMarkers = () => {
		if (this.videoStream.current) {
			return (
				<div styleName={this.state.photoTaken ? 'hide' : 'markers-ctn'}>
					<div styleName="top">
						<div styleName="markers top-left" />
						<div styleName="markers top-right" />
					</div>
					<div styleName="bottom">
						<div styleName="markers bottom-left" />
						<div styleName="markers bottom-right" />
					</div>
				</div>
			);
		}
		return <div />;
	}

	render() {
		return (
			<React.Fragment>
				<canvas
					ref={this.canvas}
					styleName={this.state.photoTaken ? 'canvas' : 'hide'}
					width={this.props.width}
					height={this.props.height}
				/>
				<video
					ref={this.videoStream}
					autoPlay
					styleName={this.state.photoTaken ? 'hide' : 'video'}
				>
					<track kind="captions" />
				</video>
				{this.state.wait && this.renderMarkers()}
			</React.Fragment>
		);
	}
}

export default CSSModules(Camera, styles, { allowMultiple: true });
