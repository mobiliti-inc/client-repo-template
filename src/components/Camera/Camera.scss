@import "../../styles/_variables.scss";

.container {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.hide {
  display: none
}

.canvas {
  display: flex;

  // @media (max-width: $mobile-width) {}
}

.video {
  padding: 0;
}

.markers-ctn {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  margin: auto;
  position: absolute;
  z-index: 100;
  top: 20%;
  bottom: 20%;
  right: 0;
  left: 0;
  width: 80%;
  height: 45%;

  @media (max-width: $mobile-width) {
    width: 85%;
    height: 23%;
  }

}


.markers {
  width: 50px;
  height: 50px;
  margin: 0;

  @media (max-width: $mobile-width) {
    width: 40px;
    height: 40px;
  }
}

.bottom {
  display: flex;
}

.bottom-left {
  border-left: 4px solid $green-check;
  border-bottom: 4px solid $green-check;
  position: absolute;
  bottom: 0;
  left: 0;
}

.top {
  display: flex;
}

.top-left {
  border-top: 4px solid $green-check;
  border-left: 4px solid $green-check;
  position: absolute;
  top: 0;
  left: 0;
}

.top-right {
  border-top: 4px solid $green-check;
  border-right: 4px solid $green-check;
  position: absolute;
  top: 0;
  right: 0;

}

.bottom {
  display: flex;
}

.bottom-right {
  border-right: 4px solid $green-check;
  border-bottom: 4px solid $green-check;
  position: absolute;
  bottom: 0;
  right: 0;
}

.button {
  display: flex;
  justify-content: center;
  position: relative;

  button {
    padding: 0;
    position: absolute;
    top: -60px;
  }

  img {
    width: 35px;
  }
}