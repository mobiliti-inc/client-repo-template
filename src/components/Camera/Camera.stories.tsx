import { storiesOf } from '@storybook/react';
import { boolean, number } from '@storybook/addon-knobs';
import * as React from 'react';
import Camera from './Camera';

storiesOf('Camera', module)
	.add('default', () => (
		<Camera
			takePhoto={boolean('takePhoto', false)}
			height={number('height', 300)}
			width={number('width', 500)}
			closeCamera={boolean('closeCamera', false)}
			onCameraAccessSuccess={() => { console.log('Camera accessed'); }}
			onCameraAccessFail={() => { console.log('Camera access failed'); }}
			allowButton={boolean('allowButton', true)}
			showMarkers={boolean('showMarkers', false)}
		/>
	)
	);
