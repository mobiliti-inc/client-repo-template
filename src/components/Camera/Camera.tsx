import * as React from 'react';
import cx from 'classnames';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCamera } from "@fortawesome/pro-light-svg-icons";

import { Button } from '../';
import './Camera.scss';

const { useState, useRef, useEffect } = React;

/*  TODO: Adjust camera markers to fit on camera
	TODO: Support picture retaking */

interface CameraProps {
	takePhoto: boolean;
	height: number;
	width: number;
	onTakingPhoto?: (...args: any[]) => any;
	closeCamera: boolean;
	onCameraAccessFail(error: any): void;
	onCameraAccessSuccess(): void;
	allowButton?: boolean;
	showMarkers?: boolean;
}

const Camera: React.FC<CameraProps> = (props) => {
	const {
		closeCamera,
		takePhoto,
		onCameraAccessSuccess,
		onCameraAccessFail,
		allowButton,
		onTakingPhoto,
		showMarkers,
		width,
		height
	} = props;

	let videoStream: any = useRef(null);
	let canvas: any = useRef(null);

	const constraints = {
		video: true
	};

	const [photoTaken, setPhotoTaken] = useState(false);
	const [localStream, setLocalStream] = useState<any>(null);

	const handleError = (error: any) => {
		onCameraAccessFail(error);
	};

	const gotStream = (stream: any) => {
		if (videoStream.current) {
			videoStream.current.srcObject = stream;
			setLocalStream(stream);
			onCameraAccessSuccess();
		}
	};

	const initiateStream = () => {
		navigator.mediaDevices
			.getUserMedia(constraints)
			.then(stream => gotStream(stream))
			.catch(handleError);
	};

	const hasGetUserMedia = () =>
		!!(navigator.mediaDevices && navigator.mediaDevices.getUserMedia);

	const stopTracks = () => {
		if (localStream) {
			const tracks = localStream.getTracks();
			videoStream.current.srcObject = null;
			tracks.forEach((track: any) => {
				track.stop();
			});
		}
	};

	const captureImage = () => {
		setPhotoTaken(true);
		const context = canvas.current.getContext('2d');
		context.imageSmoothingEnabled = false;
		context.height = height;
		context.width = width;
		context.drawImage(videoStream.current, 0, 0, 300, 150);
		const image = canvas.current.toDataURL('image/jpeg', 0.5);
		stopTracks();
		return onTakingPhoto && onTakingPhoto(image);
	};

	const retake = () => {
		setPhotoTaken(false);
		initiateStream();
	};

	useEffect(() => {
		if (photoTaken) {
			stopTracks();
		}
		if (takePhoto) {
			if (photoTaken && !closeCamera) {
				retake();
			}
		}
	}, [closeCamera, takePhoto]);

	useEffect(() => {
		if (hasGetUserMedia()) {
			initiateStream();
		} else {
			const error = {
				error: 'Unable to access user media'
			};
			onCameraAccessFail(error);
		}
	}, []);

	const renderMarkers = () => {
		if (videoStream.current) {
			return (
				<div styleName={photoTaken ? 'hide' : 'markers-ctn'}>
					<div styleName={'top'}>
						<div styleName={cx('markers', 'top-left')} />
						<div styleName={cx('markers', 'top-right')} />
					</div>
					<div styleName={'bottom'}>
						<div styleName={cx('markers', 'bottom-left')} />
						<div styleName={cx('markers', 'bottom-right')} />
					</div>
				</div>
			);
		}
		return <div />;
	};

	return (
		<div styleName="container">
			<canvas
				ref={canvas}
				styleName={photoTaken ? 'canvas' : 'hide'}
				style={{ width: `${width}px`, height: `${height}px` }}
			/>
			<video
				ref={videoStream}
				autoPlay
				styleName='video'
				style={{ width: `${width}px`, height: `${height}px` }}
			/>
			{allowButton && !photoTaken && (
				<div styleName="button">
					<FontAwesomeIcon icon={faCamera} />
					<Button shape="round" ripple onClick={() => captureImage()}>[]</Button>
				</div>
			)}
			{showMarkers && renderMarkers()}
		</div>
	);
};

export default Camera;
