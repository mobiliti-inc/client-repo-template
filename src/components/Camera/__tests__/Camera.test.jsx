import Camera from '../Camera';

describe('Camera', () => {
	it('renders', () => {
		const tree = shallow(
			<Camera
				onTakingPhoto={() => { }}
				takePhoto
				height={300}
				width={500}
				closeCamera={false}
			/>);
		matchSnapshot(tree);
	});

	it('renders video element so one can take photo ', () => {
		const subject = shallow(
			<Camera
				onTakingPhoto={() => { }}
				takePhoto
				height={300}
				width={500}
				closeCamera={false}
			/>);
		matchSnapshot(subject);
		expect(subject.html()).toContain(
			'<canvas width="500" height="300" class="hide"></canvas><video autoplay="" class="video"><track kind="captions"/></video>');
	});
});
