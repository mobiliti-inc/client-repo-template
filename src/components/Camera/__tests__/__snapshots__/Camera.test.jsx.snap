// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Camera renders 1`] = `
<React.Fragment>
  <canvas
    className="hide"
    height={300}
    width={500}
  />
  <video
    autoPlay={true}
    className="video"
  >
    <track
      kind="captions"
    />
  </video>
</React.Fragment>
`;

exports[`Camera renders video element so one can take photo  1`] = `
<React.Fragment>
  <canvas
    className="hide"
    height={300}
    width={500}
  />
  <video
    autoPlay={true}
    className="video"
  >
    <track
      kind="captions"
    />
  </video>
</React.Fragment>
`;
